earthmover/earthmover.py

```diff
@@ -383,3 +390,155 @@ def test(self, tests_dir: str):
             if not _expected_df.equals(_outputted_df):
                 self.logger.critical(f"Test output `{_outputted_file}` does not match expected output.")
                 exit(1)
+
+
+    def deps(self):
+        """
+        Installs all packages specified in the config file and any nested packages.
+        :return:
+        """
+        self.build_root_package_graph()
+
+        # Check that at least one package is defined
+        if all(False for _ in self.package_graph.successors('root')):
+            self.error_handler.throw("No packages have been defined!")
+
+        # Install each package (and any nested sub-packages) into the packages directory
+        self.build_package_graph(root_node='root', package_subgraph=self.package_graph, packages_dir=self.packages_dir, install=True)
+
+
+    def merge_packages(self):
+        """
+        Traverses the packages graph, merging yaml config from successors into predecessors.
+        Saves the final result as the instance user_configs.
+        :return:
+        """
+        self.build_root_package_graph()
+
+        # If the yaml file doesn't include packages, no need to alter
+        if all(False for _ in self.package_graph.successors('root')):
+            return
+        
+        self.build_package_graph(root_node='root', package_subgraph=self.package_graph, packages_dir=self.packages_dir, install=False)
+
+        # Merge each package yaml into the predecessor yaml, storing the result in the predecessor
+        # Post-order traversal ensures the correct hierarchy of merges
+        for package_name in nx.dfs_postorder_nodes(self.package_graph):
+            package_node = self.package_graph.nodes[package_name]
+
+            for predecessor_name in self.package_graph.predecessors(package_name): # more elegant way to do this? we know each node will only have one predecessor
+                predecessor_node = self.package_graph.nodes[predecessor_name]
+                merged_yaml = self.merge_dicts(package_node['package'].package_config, predecessor_node['package'].package_config)
+                predecessor_node['package'].package_config = merged_yaml
+
+        # Overwrite with completed merged yaml  
+        self.user_configs = self.package_graph.nodes['root']['package'].package_config
+        
+        # Output merged yaml file
+        # TODO: make this prettier
+        with open("./merged_earthmover.yml", "w") as f:
+            json.dump(self.user_configs, f, ensure_ascii=False, indent=4)

```

Comment:
@sleblanc23 I was playing around with this locally and landed on the following, less-pretty code, but it produces nice, pretty YAML:
```python
with open("./merged_earthmover.yml", "w") as f:
            yaml.safe_dump(json.loads(json.dumps(self.user_configs, ensure_ascii=True)), f, default_flow_style=False)
```
Working from the inside out,
* `json.dumps()` produces the JSON-formatted object as a string, but crucially _without the line number annotations_ earthmover adds when parsing YAML (for debugging purposes)
* `json.loads()` turns the string back into a Python object
* `yaml.safe_dump()` turns the Python object into YAML and writes it to the file

Maybe there's a nicer/better way to do this, just thought I'd share in case this is helpful.

Reply:
Thank you, Tom! I'd landed on something similar but not quite as pretty :) Appreciate you sharing